parser-typechecker/src/Unison/Typechecker/Context.hs

```diff
@@ -173,6 +180,14 @@ instance Monad (Result v loc) where
   CompilerBug bug es is >>= _ = CompilerBug bug es is
   {-# INLINE (>>=) #-}
 
+instance MonadFix (Result v loc) where
+  mfix f =
+    let res = f theA
+        theA = case res of
+          Success _ a -> a
+          _ -> error "mfix Result: forced an unsuccessful value"
```

Comment:
At a glance this seems like a potential issue, isn't it possible that monadfix could try to force something that has an error, and in  hitting this case would swallow the meaningful error with this one?

Maybe it's not possible right now because we only pass things that typecheck to the pattern-match-checker?

Reply:
It is true in general when using value recursion that you could force the the thunk before it is resolved and cause your program to hang, in this case forcing the thunk when you don't return a `Success` would be an error. This is also true in the `MonadFix` instances for `Maybe`, `Either`, and such.

The danger of doing so is in an `mfix` call, or recursive do block, as that's when you have this thunk that will be a bomb if the result isn't a `Success`. There isn't an `mfix` call that spans the typechecker and pattern match coverage checker or something like that though.

Reply:
This example might make what I'm saying clearer:

`goodExample` replaces two ints in a `Map` with a thunk that will resolve to their sum (avoiding a second pass of inserts). `forceLeftExample` and `loopExample` illustrate erroneous ways to use `mfix`.
```haskell
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TupleSections #-}

import Control.Exception
import Control.Monad.Fix
import Data.Functor.Compose
import Data.Map (Map)
import qualified Data.Map as Map

main :: IO ()
main = do
  print (goodExample (Map.fromList [(A, 2)]))
  print (goodExample (Map.fromList [(A, 2), (B, 3)]))
  print (forceLeftExample (Map.fromList [(A, 2), (B, 3)])) `catch` \(e :: SomeException) -> do
    putStrLn ("caught exception: " <> show e)
  print (loopExample (Map.fromList [(A, 2), (B, 3)]))

goodExample :: Map Key Int -> Either String (Map Key Int)
goodExample m = snd <$> mfix summedMap
  where
    summedMap :: (Int, Map Key Int) -> Either String (Int, Map Key Int)
    summedMap ~(theSum, _) = do
      (a, m1) <- replaceAndReturnOriginal A theSum m
      (b, m2) <- replaceAndReturnOriginal B theSum m1
      pure (a + b, m2)

replaceAndReturnOriginal :: forall k v. (Show k, Ord k) => k -> v -> Map k v -> Either String (v, Map k v)
replaceAndReturnOriginal k v n =
  let alterFunc :: Maybe v -> Compose (Either String) ((,) v) (Maybe v)
      alterFunc x = Compose $ (,Just v) <$> toEither x

      toEither :: Maybe v -> Either String v
      toEither = \case
        Nothing -> Left ("Missing key: " <> show k)
        Just x -> Right x
   in getCompose $ Map.alterF alterFunc k n

loopExample :: Map Key Int -> Either String (Map Key Int)
loopExample m = snd <$> mfix summedMap
  where
    summedMap :: (Int, Map Key Int) -> Either String (Int, Map Key Int)
    summedMap ~(theSum, _) = do
      (a, m1) <- replaceAndReturnOriginal A theSum m
      (b, m2) <- theSum `seq` replaceAndReturnOriginal B theSum m1 -- we force theSum while computing theSum
      pure (a + b, m2)

forceLeftExample :: Map Key Int -> Either String (Map Key Int)
forceLeftExample m = snd <$> mfix summedMap
  where
    summedMap :: (Int, Map Key Int) -> Either String (Int, Map Key Int)
    summedMap bomb = do
      Left (show bomb) -- we smuggled our thunk out in a Left, but this will error due to the MonadFix instance

data Key = A | B deriving (Eq, Ord, Show)
```

The output is:
```
Left "Missing key: B"
Right (fromList [(A,5),(B,5)])
caught exception: mfix Either: Left
test: <<loop>>
```